generator/src/main/java/com/scottlogic/deg/generator/reducer/ConstraintReducer.java

```diff
@@ -82,16 +82,15 @@ public ConstraintReducer(
     }
 
     private Optional<FieldSpec> getRootFieldSpec(Iterable<AtomicConstraint> rootConstraints) {
-        final Stream<FieldSpec> rootConstraintsStream =
+        final Stream<Optional<FieldSpec>> rootConstraintsStream =
             StreamSupport
                 .stream(rootConstraints.spliterator(), false)
-                .map(fieldSpecFactory::construct);
+                .map(fieldSpecFactory::construct)
+                .map(Optional::of);
 
         return rootConstraintsStream
             .reduce(
                 Optional.of(FieldSpec.Empty),
-                (optAcc, next) ->
-                    optAcc.flatMap(acc -> fieldSpecMerger.merge(acc, next)),
                 (optAcc1, optAcc2) -> optAcc1.flatMap(
                     acc1 -> optAcc2.flatMap(
```

Comment:
we know that optAcc2 will always be present, we can skip this 2nd flatmap

also, would be nice to rename from acc to spec

Reply:
Turns out we don't know the second will always be present, due to reduce not guaranteeing an ordering on the reduce. We can get into a situation where either the first or the second have no more values. At this point the result will be `Optional.empty` and no more values can be generated.

Reply:
are you sure about this? is there docs for it?

Reply:
Yep, on the method `java.util.stream.Stream#reduce(T, java.util.function.BinaryOperator<T>)`